### PR TITLE
Cleanup | Consolidate connection capability metadata

### PR DESCRIPTION
**Pull request by edwardneal**
## Description

This PR emerges in part from discussion with @paulmedynski on #3858. We noticed that the `json` datatype wouldn't appear when running `SqlConnection.GetSchema("DataTypes")` against an Azure SQL instance. This is because of an underlying design choice in SqlMetaDataFactory: it only uses the SQL Server version number to control whether specific queries are used and specific records are returned. This isn't compatible with Azure SQL (which always returns a version of 12.x.) To fix this, we need SqlMetaDataFactory to be able to make decisions based upon the server capabilities, whether determined by the version, the presence of a FEATUREEXTACK token, or the contents of one of those tokens' values.

In this PR, we introduce a new type, `ConnectionCapabilities`. This class takes on the responsibility of hosting the various feature flags as an object which SqlConnectionInternal, SqlMetaDataFactory and TdsParser can interrogate to check for feature availability. For good measure, I then plumb this object through to SqlMetaDataFactory. A subsequent PR can work out the best way to implement filtering.

Other points of errata:
* Moving the LOGINACK capability/version interrogation into the new type means that we no longer need to keep the original `SqlLoginAck` instance as a field on SqlConnectionInternal, which means that it never leaves the scope it was created in. I've turned it into a readonly ref struct.
* 229a04de860d9b76e49cc3aeda4553326bd2fd80 enables SqlClient to build under the Release configuration. This was necessary to get the benchmarking to work.
* Benchmarking confirms that the CPU time, number of Gen0 GCs and memory usage remains identical (or very slightly better - slightly fewer Gen0 collections) between this branch and main.

This is a comparatively large PR, so I've tried to make it practical to move commit-by-commit.

## Issues

Builds on a conversation in #3858.
Prerequisite to #3833.

## Testing

This is covered by existing tests.

### CODE REVIEW

**Comment by edwardneal** [comment]
This PR is not stale.

**Comment by paulmedynski** [comment]
@edwardneal - Can you resolve the conflicts?  I'll kick a new PR run once that's done.

**Comment by paulmedynski** [comment]
@edwardneal - The PR pipeline issues on main have been fixed.  Can you re-merge this and your owht Cleanup PRs?  I'll get the pipelines started on all of them.  Sorry!

**Comment by edwardneal** [comment]
Thanks @paulmedynski. I've just merged main into this PR. ~I'll merge into the rest of the PRs over the next few days - if you're waiting on any in particular for GA then just let me know and I'll prioritise those.~

I've just merged main into the other two cleanup PRs.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    /// If true, this connection is to an Azure SQL instance. This is determined
+    /// by the receipt of a FEATUREEXTACK token of value <c>0x08</c>.
+    /// </summary>
+    public bool IsAzureSql { get; private set; }
+
+    /// <summary>
+    /// Indicates support for user-defined CLR types (up to a length of 8000
+    /// bytes.) This was introduced in SQL Server 2005.
+    /// </summary>
+    public bool UserDefinedTypes => true;
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs:80]: Note to self: expecting to see these get cleaned up

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+                    && featureData[0] <= TdsEnums.MAX_SUPPORTED_VECTOR_VERSION;
+                break;
+
+            case TdsEnums.FEATUREEXT_JSONSUPPORT:
+                // Feature data is comprised of a single byte which specifies the version of the JSON
+                // type which is available.
+                JsonType = !featureData.IsEmpty && featureData[0] != 0x00
+                    && featureData[0] <= TdsEnums.MAX_SUPPORTED_JSON_VERSION;
+                break;
+
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs]: missing a case for enhanced routing

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs [hunk 3]
```diff
@@ -1273,16 +1273,12 @@ internal void OnError(SqlException exception, bool breakConnection, Action<Actio
             }
         }
 
-        // @TODO: This feature is *far* too big, and has the same issues as the above OnEnvChange
-        // @TODO: Consider individual callbacks for the supported features and perhaps an interface of feature callbacks. Or registering with the parser what features are handleable.
+        internal bool ShouldProcessFeatureExtAck(byte featureId) =>
+            RoutingInfo is null || featureId is TdsEnums.FEATUREEXT_SQLDNSCACHING;
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs]: needs updating for enhanced routing

**Review by mdaigle — CHANGES_REQUESTED** [review]
I'm ok to proceed with this, but something about it rubs me the wrong way, and I think it's that there's no way to know when the class is fully populated. We have two separate "process" methods that need to be called before you can read from the class's properties. If you do read before it's fully populated you'll get different values out. That said, the same critique applies to what we have now. I think this is a step in the right direction, but want to keep iterating on how to have strong guardrails during token processing so that it's hard to make mistakes.

Will take another look when the enhanced routing support is incorporated!

**Comment by edwardneal** [comment]
Thanks @mdaigle - I've moved the enhanced routing detection.

I think part of the issue here is that the LOGINACK and the FEATUREEXTACK tokens are treated independently in TdsParser. The logic to read these tokens from the wire is blended with the parsing and processing, and that flexibility has meant that anything downstream has to assume it can receive those tokens in any order; it never knows precisely when the item has finished processing, so we can't strictly guarantee that anything will be read-only.

If we could get to grips with the message sequencing, start to become more stringent about the order we expect to receive token streams from the server and start to factor that parsing out from TdsParser, I'd quite like to have some form of "capability builder" class which accepts a single LOGINACK token, any number of FEATUREEXTACK tokens and constructs a static ConnectionCapabilities instance for downstream call sites.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs (hunk 3 above)
Comment (mdaigle) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs]: This check needs to include enhanced routing as an "or" condition. We get envchange tokens before feature ext tokens, so if we don't have a special case for it we won't process the enhanced routing ack.

Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs]: This condition is already in the latest version of the file (L1265.)

**Comment by paulmedynski** [comment]
@edwardneal - New conflicts here.

**Comment by edwardneal** [comment]
Thanks @paulmedynski - just merged.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionFactory.cs [hunk 4]
```diff
@@ -757,11 +757,12 @@ private static DbConnectionPoolGroupOptions CreateConnectionPoolGroupOptions(Sql
...
         {
             Debug.Assert(internalConnection is not null, "internalConnection may not be null.");
+            Debug.Assert(internalConnection as SqlConnectionInternal is not null, "innerConnection must be a SqlConnectionInternal.");
 
             Stream xmlStream = Assembly.GetExecutingAssembly().GetManifestResourceStream("Microsoft.Data.SqlClient.SqlMetaData.xml");
             Debug.Assert(xmlStream is not null, $"{nameof(xmlStream)} may not be null.");
-            
-            return new SqlMetaDataFactory(xmlStream, internalConnection.ServerVersion);
+
+            return new SqlMetaDataFactory(xmlStream, ((SqlConnectionInternal)internalConnection).Parser.Capabilities);
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionFactory.cs]: Can this be pushed up to some level where it's clear it will succeed? Then we can scope down the param type.

**Review by mdaigle — COMMENTED** [review]
On second review, I don't think I'm comfortable with shifting so much parsing logic around without a long term plan or vision in place. I think the piece that clearly makes sense is a record to store and pass capabilities. Can we reduce the changes to just that?

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 5]
```diff
@@ -3855,18 +3845,23 @@ private TdsOperationStatus TryProcessFeatureExtAck(TdsParserStateObject stateObj
...
+                        Capabilities.ProcessFeatureExtAck(featureId, data);
+                        _connHandler.OnFeatureExtAck(featureId, data);
+                    }
                 }
             } while (featureId != TdsEnums.FEATUREEXT_TERMINATOR);
 
             // Write to DNS Cache or clean up DNS Cache for TCP protocol
             bool ret = false;
-            if (_connHandler._cleanSQLDNSCaching)
+            if (!Capabilities.DnsCaching)
```
Comment (priyankatiwari08) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs]: Behavior change worth confirming: this was previously if (_connHandler._cleanSQLDNSCaching). That field defaults to false and was only set to true in OnFeatureExtAck when the server sent SQLDNSCACHING with data[0] == 0. So in main, connections to servers that don't send SQLDNSCACHING at all (i.e., non-Azure) did not call DeleteDNSInfo. With if (!Capabilities.DnsCaching) the inverse is true — those connections will now hit DeleteDNSInfo on every login. Intentional?

**Review by priyankatiwari08 — COMMENTED** [review]
Agree with @mdaigle's ask to reduce scope to just the capability record + SqlMetaDataFactory plumbing and defer the parsing relocation. Left one inline on TdsParser.cs about a DNS-caching cleanup behavior change in that relocated code — worth confirming before any of it lands.

**Comment by edwardneal** [comment]
Thanks both. I've pared this back, so `ConnectionCapabilities` is just a resettable POCO now. This exposed that the `SqlLoginAck` class was no longer necessary, so I've removed it.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionFactory.cs (hunk 4 above)
Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionFactory.cs]: I've pushed this up to `DbConnectionInternal`, returning null by default and matching the throwing/non-throwing behaviour on all derived classes.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 5 above)
Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs]: Yes, correct - unintentional and removed. Thanks for pointing that out.

**Comment by mdaigle** [comment]
Thanks, I will take another look on Monday.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 6]
```diff
@@ -4317,50 +4298,23 @@ private TdsOperationStatus TryProcessLoginAck(TdsParserStateObject stateObj, out
...
 
-            // Server responds:
-            // 0x07000000 -> 7.0         // Notice server response format is different for bwd compat
-            // 0x07010000 -> 2000 RTM     // Notice server response format is different for bwd compat
-            // 0x71000001 -> 2000 SP1
-            // 0x72xx0002 -> 2005 RTM
-            // information provided by S. Ashwin
-            switch (majorMinor)
+            if (tdsVersion is not TdsEnums.SQL2005_VERSION
+                and not TdsEnums.SQL2008_VERSION
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs]: Can you take a close look at this and double check that this is safe? The masking behavior is making me think that there are a range of accepted server-side values.

**Review by mdaigle — COMMENTED** [review]
Looking good. just feeling hesitant about the changes to server-side version calculation.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 6 above)
Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs]: I've re-checked. That first impression was part of the reason why I've left it in here post-simplification - the `switch` is a little odd.

`majorMinor` and `increment` are calculated as below in TdsParser (lines 4321-4322):
```cs
// Bits 0-15 and 24-31
majorMinor = [TdsVersion] & 0xFF00FFFF;

// Shift right by 16 bits and take lower 8 bits. Equivalent maths is below.
increment = ([TdsVersion] >> 16) & 0xFF;
increment = ([TdsVersion] & 0x00FF0000) >> 16;
```

There are four cases: one per combination of `majorMinor`. Each case has one valid `increment` value, and anything outside of these cases fails. They're present in TdsParser (lines 4332-4360.)

|  | SQL 2005 | SQL 2008 | SQL 2012 / TDS 7.x | TDS 8 |
| - | - | - | - | - |
| `majorMinor` expression | `(SQL2005_MAJOR << 24) OR SQL2005_RTM_MINOR` | `(SQL2008_MAJOR << 24) OR SQL2008_MINOR` | `(SQL2012_MAJOR << 24) OR SQL2012_MINOR` | `(TDS8_MAJOR << 24) OR TDS8_MINOR` |
| Major version constant | 0x72 | 0x73 | 0x74 | 0x08 |
| Minor version constant | 0x0002 | 0x0003 | 0x0004 | 0x0000 |
| Resultant `majorMinor` | 0x72_00_0002 | 0x73_00_0003 | 0x74_00_0004 | 0x08_00_0000 |
| Acceptable `increment` | `SQL2005_INCREMENT` | `SQL2008_INCREMENT` | `SQL2012_INCREMENT` | `TDS8_INCREMENT` |
| Resultant `increment` | 0x09 | 0x0B | 0x00 | 0x00 |

Since both `majorMinor` and `increment` come from the same `[TdsVersion]` value, we can OR them together and perform the comparison in one step, resulting in four cases:
* SQL 2005: 0x72_09_0002 (`SQL2005_VERSION`)
* SQL 2008: 0x73_0B_0003 (`SQL2008_VERSION`)
* SQL 2012: 0x74_00_0004 (`TDS7X_VERSION`)
* TDS 8: 0x08_00_0000 (`TDS80_VERSION`)

This corresponds to the values specified in MS-TDS for the LOGINACK token stream for every value besides TDS 8 - [Appendix A, <72>](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/135d0ebe-5c4c-4a94-99bf-1811eccb9f4a#Appendix_A_72).

There's also prior art in the tediousjs version mapping [here](https://github.com/tediousjs/tedious/blob/master/src/tds-versions.ts) and in mssql-jdbc [here](https://github.com/microsoft/mssql-jdbc/blob/main/.github/instructions/glossary.instructions.md#version-specific-terms).

This `switch` also contains the logic determining whether the remote server is SQL 2022 or newer / SQL 2012 or newer / SQL 2008 R2 or newer. I've moved this logic to `ConnectionCapabilities`, lines 54-69.

Comment (mdaigle) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs]: Makes sense, and the new code is much cleaner. Can we refactor this logic into a static method and add some unit testing to prove equivalency (copy old code into test and compare)? I think I need to see that to feel comfortable with this. Plus, it will give a good spot to verify version parsing if/when we add a new version in the future.

Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs]: I've just shuffled the validation logic and added a unit test in `AdapterUtilTest`.

**Review by mdaigle — APPROVED** [review]
Looks good. Thanks!

**Comment by mdaigle** [comment]
@priyankatiwari08 this is ready for another review

**Review by mdaigle — DISMISSED** [review]

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    /// </summary>
+    public bool Is2012OrNewer =>
+        Is2022OrNewer || TdsVersion == TdsEnums.TDS7X_VERSION;
+
+    /// <summary>
+    /// If true (as determined by the value of <see cref="TdsVersion"/>)
+    /// then the connection is to SQL Server 2022 or newer.
+    /// </summary>
+    public bool Is2022OrNewer =>
+        TdsVersion == TdsEnums.TDS80_VERSION;
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs]: Is it possible that we chose to negotiate TDS 7.4 when connecting to SqlServer 2022+ (i.e. for TLS version or session encryption reasons)?

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    /// If true, this connection is to an Azure SQL instance. This is determined
+    /// by the receipt of a FEATUREEXTACK token of value <c>0x08</c>.
+    /// </summary>
+    public bool IsAzureSql { get; set; }
+
+    /// <summary>
+    /// Indicates support for user-defined CLR types (up to a length of 8000
+    /// bytes.) This was introduced in SQL Server 2005.
+    /// </summary>
+    public bool UserDefinedTypes => true;
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs:80]: Thank you for not prepending `Is` or `Has` to these 😄 .  I would even be happy with `AzureSql` and `AtLeastSql2022`, but am fine with those as-is.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+#nullable enable
+
+namespace Microsoft.Data.SqlClient;
+
+/// <summary>
+/// Describes the capabilities and related information (such as the
+/// reported server version and TDS version) of the connection.
+/// </summary>
+internal sealed class ConnectionCapabilities
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs:17]: Are these properties immutable once the session is established?  If so, it would be nice for instances of this class to be immutable as well.  I suspect I'll see you creating it default and then filling in the flags as we process the login flow, so it is indeed convenient for that bit-by-bit accumulation.  I just worry about accidental changes post-login.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    /// <summary>
+    /// If column encryption is enabled, the type of enclave reported by the server. This
+    /// was introduced in SQL Server 2019, and is only available if a FEATUREEXTACK token
+    /// of value <c>0x04</c> is received, and the resultant <see cref="ColumnEncryptionVersion"/>
+    /// is <c>2</c> or <c>3</c>.
+    /// </summary>
+    public string? ColumnEncryptionEnclaveType { get; set; }
+
+    /// <summary>
+    /// Returns the capability records to unset values.
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs:199]: Intereseting - I was expecting this to live/die per-TDS-session+connection.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs [hunk 11]
```diff
@@ -1523,19 +1466,12 @@ internal void OnFeatureExtAck(int featureId, byte[] data)
...
-                    _parser.IsColumnEncryptionSupported = true;
-                    _parser.TceVersionSupported = _tceVersionSupported;
-                    _parser.AreEnclaveRetriesSupported = _tceVersionSupported == 3;
+                    Capabilities.ColumnEncryptionVersion = supportedTceVersion;
 
                     if (data.Length > 1)
                     {
                         // Extract the type of enclave being used by the server.
-                        _parser.EnclaveType = Encoding.Unicode.GetString(data, 2, data.Length - 2);
+                        Capabilities.ColumnEncryptionEnclaveType = Encoding.Unicode.GetString(data, 2, data.Length - 2);
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs:1471]: One possible behavioral change in OnFeatureExtAck for FEATUREEXT_TCE: enclave-related state is now derived from TceVersionSupported (via Capabilities.ColumnEncryptionVersion) instead of being set independently.

In particular, previously enclave type assignment was effectively driven by payload length, while now enclave semantics are tied to CE version (v2/v3). That seems correct, but it does change edge-case behavior for malformed/nonstandard payloads (e.g., v1 with extra enclave bytes).

Could we add/point to a test that locks this in explicitly?
- v1: column encryption supported, no enclave semantics
- v2: enclave type present/usable
- v3: enclave retries supported

That would make the “no regression for valid inputs, stricter handling for invalid/legacy edge cases” intent very clear.

Looking for commentary on this before any changes (if at all).

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 12]
```diff
@@ -402,8 +388,8 @@ bool withFailover
             _connHandler = connHandler;
             _loginWithFailover = withFailover;
 
-            // Clean up IsSQLDNSCachingSupported flag from previous status
-            _connHandler.IsSQLDNSCachingSupported = false;
+            // Clean up all server capabilities from previous status.
+            Capabilities.Reset();
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:392]: Why not `Capabilities = new()`?  Is this trying to avoid allocations?  Are Capabilities small enough to put on the stack (i.e. a struct)?

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 13]
```diff
@@ -273,13 +271,7 @@ internal EncryptionOptions EncryptionOptions
...
         }
 
-        internal bool Is2008OrNewer
-        {
-            get
-            {
-                return _is2008;
-            }
-        }
+        internal bool Is2008OrNewer => Capabilities.Is2008R2OrNewer;
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:274]: Can we remove this and let callers use TdsParser.Capabilities instead?  They have the same scope/visibility.

**Review by paulmedynski — CHANGES_REQUESTED** [review]
Great changes overall and a good step towards encapsulating concerns in this old spaghetti code!  A few questions/comments for your consideration.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs (hunk 9 above)
Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs:17]: You're correct, it'll need to be filled in as the login flow proceeds. It's pretty ugly, but it flows downstream from the wider design of the parser. SqlClient just processes FEATUREEXTACK token streams when it receives them. It could (in theory) receive a FEATUREEXTACK token stream from a broken server at any stage, and this would be processed without a hitch.

There's also the case of a client redirection to another server to consider. In such a situation, we might end up needing to make changes as the result of routing following the receipt of  a LOGINACK token stream.

I suspect that actually fixing this (and hardening against noncompliant/malicious servers) might involve peeling some of this processing away from TdsParser. That'd be a good idea, but it needs far more thought and test coverage of the login flows first.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 13 above)
Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:274]: We can, but there are dozens of these call sites - sometimes for each specific capability. I'd like to replace them all at once in a follow-up.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 12 above)
Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:392]: The first reason is to avoid allocations. We've gained a new object instantiation in `ConnectionCapabilities`, but we're no longer creating one in `SqlLoginAck` so it roughly balances out.

Besides allocations though, this comes closest to mirroring the existing model of resetting existing capability flags.

We can't put Capabilities on the stack - it's modified by `TdsParser.TryProcessFeatureExtAck` (and from there, `SqlConnectionInternal.OnFeatureExtAck`) and thus needs to be a property on `TdsParser`.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs (hunk 11 above)
Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs:1471]: Could you clarify this please? Lines 1458-1467 already guaranteed `supportedTceVersion` is strictly greater than  zero and less than or equal to `TdsEnums.MAX_SUPPORTED_TCE_VERSION (0x03)` - so by the time we exit that if-condition and start setting `Capabilities.ColumnEncryptionVersion`, we already have a guarantee that the value we give to it will be 1, 2 or 3.

With that in mind:
* `ColumnEncryptionVersion` replaces `TceVersionSupported`.
* `IsColumnEncryptionSupported` makes more sense - it just tests to see if `ColumnEncryptionVersion` is greater than zero.
* `AreEnclaveRetriesSupported` is currently set to `true` if `TceVersionSupported == 3`. This is replaced by an expression of `ColumnEncryptionVersion >= 3` (essentially the same thing, given that we're constraining the TCE versions we support as per above.)

`ColumnEncryptionEnclaveType` continues to be set based upon token payload length.

As far as commentary goes, TCE is definitely a divergence from the TDS spec ([link](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/2eb82f8e-11f0-46dc-b42d-27302fa4701a)). I just didn't want to make a correctness change at the same time as a refactor.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs (hunk 7 above)
Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs]: Absolutely. This is just a drop-in replacement for the existing `_is2022` variable in TdsParser, which is itself used solely to make sure that TDS 8.0 doesn't break a few features which are gated upon SQL Server 2008 support (namely, UDTs and the XML data type.)

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 13 above)
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:274]: Ok look forward to it!

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs (hunk 11 above)
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs:1471]: Ok agreed - let's not make any behaviour changes.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs (hunk 7 above)
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs]: Ahh I see - so `_is2022` was also misleadingly named.  It also really meant "We negotiated TDS 8.x".  I guess all 3 of these properties suffers from confusing naming, but they retain the naming from the old `TdsParser` fields.

A connection to SQL 2025 using TDS 7.x will have `Is2012OrNewer` true and `Is2022OrNewer` false, which is making my head hurt 😄 

Could you update the XML docs to be super clear that these are solely a reflection of the negoitated TDS version, and don't reflect the server version despite their names?  Perhaps even consider renaming these to `NegotiatedTds80` and similar.

Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/ConnectionCapabilities.cs]: The field's logic was a little tangled. I've dug into a little detail though, and we never actually need to discriminate based upon having negotiated a TDS 8.0 connection - `_is2022` was never necessary.

We need to care about the remote server running SQL Server 2012 or later, but SQL 2012 onwards always used the same pair of version fields: one for TDS 7.x, another for TDS 8.0.

With that in mind, I've removed the `Is2022OrNewer` property and changed `Is2012OrNewer` to match those semantics.

**Review by paulmedynski — APPROVED** [review]

**Review by mdaigle — APPROVED** [review]
